### ISSUE TITLE
Cones Templates: Increased Support

### ISSUE THREAD

**Issue by dorpond** [RPTools/maptool#1276]
**Is your feature request related to a problem? Please describe.**
Current cone template does not support D&D5e.

**Describe the solution you'd like**
Our current cone template was built around 3e, so do we update that to 5e or keep it? Jamz said that Pathfinder uses the existing template, so maybe adding a second cone template, much like we currently do with lines and spheres would be all we need until we reinvent the whole template system.

**Describe alternatives you've considered**
If the toolbar is getting too loaded up with templates, perhaps we should have all the templates in campaign properties for us to pick from, and only those show up on the toolbar?

**Additional context**
I just think MT really needs to support the top gaming systems being played out there, or else people will go to another VTT that does.

In the attached jpg, the current cone templates are on the left and the way they should look for 5e is on the right.
![Maptool Cones](https://user-images.githubusercontent.com/31480384/74460164-78448100-4e5a-11ea-86d8-929845d23992.JPG)

**Comment by adyoungblood** [comment]
This issue's been getting to me and my group, as well as @Einfachllya#9105 on the Discord, who writes:

> So to summarize it : 
In DnD 5e, an X-foot cone originating from point P means that the cone's apex is at P and that a line segment of length X feet should first be drawn from point P in a direction of the player's choice. Then at each point Q along this line segment, draw a line segment perpendicular to the X-foot line at Q. This perpendicular line segment has length equal to |PQ| and is centered on Q.

> In simpler terms, the diameter of the cross-section of the cone taken at Q should be equal to |PQ|. Or, in plainer but less precise terms, the length along the axis from the apex at any given point should equal the diameter of the cone at that given point.

> In Maptool, "X-foot-cone originating from point P" is being interpreted as a cone whose apex is P and whose slant height is X feet. Since the slant height of a cone is always greater than the height of a cone, this means that an "X-foot-cone" in maptool is smaller or shorter than a cone in official DnD 5e rules. Further, using the slant height as the parameter that defines the cone means that the cone is largest when the axis is on a 45-degree diagonal from point P and shortest when the cone axis is parallel to the coordinate axes. This means maptool will interpret "X-foot-cone" as being of different sizes, depending on the angle at which the player aims the cone.

I'm thinking that a more flexible template system could be a great selling point for Maptool; at the moment, most popular VTTs have a limited selection of templates and no more. Perhaps it could work by letting the user define arbitrary shapes (via formulas, maybe) that include a point of origin, and then mapping those onto grid tiles on the fly. There could be a setting for how overlap should work (does it count if the center is in the area? Half the square is in the area? Etc.)

**Comment by Azhrei** [comment]
I don’t see a practical way of creating user-defined templates.  For example, it’s not practical for the user to specify which cells apply to a template since templates could be almost any size. Using formulas might work, but we’d need someone who did better in geometry and trig than I did, that’s for sure!

I would definitely like to see templates grouped by game system, though.  The GM would choose a “personality” for the campaign and that would be reflected in the default set of token properties, default lights/sights, and default set of templates.  It likely would also need to be parameterized based on the grid style (square or too-hex or side-hex).

**Comment by adyoungblood** [comment]
Palm-to-forehead moment, instead of making users define templates with formulas (might even require parametric functions and the like...), add a simple GUI to let them build their own templates out of simple shapes! Put a semicircle at the end of a triangle and you've got a 3e cone, or something close to it. The user could choose the point of origin, then draw an arrow from there to define a unit length. I can make a mockup if needed.

Of course, that's a whole other menu to build and support, and I haven't the faintest clue how to build one in Swing, much less maintain it 😅 Maybe just a few more templates for popular game systems would be a better fix for now.

**Comment by mal-w** [comment]
It still needs some work, but I have something partially working for this:
![image](https://github.com/RPTools/maptool/assets/60550827/47dd6439-1c0b-470c-a6f1-f82aff7873ab)

You can see the overlay (triangle) matches @adyoungblood 's quoted description for 5e.

I still have to figure out a couple things:

1. The way I've implemented it, basically any square that has X percent of it covered by the triangle is considered in the AOE. I don't know what to set this number to lol, or if it's to be a setting in the UI.. what to do with that? Currently it's just set to 50% of the square covered, but this is not based on any rule that I know of.
2. It's a bit finicky still with where you place your mouse to get it to work.. I'm pretty sure i can smoothe that out

**Comment by mal-w** [comment]
Discussion

https://discord.com/channels/296230822262865920/494598206626463755/1189792881179893801

We're unclear if the change proposed in [4588](https://github.com/RPTools/maptool/pull/4588) actually resolves this issue since the approach taken was to use the [template method for areas of effect on a grid](https://5e.tools/book.html#xge,2,template%20method,0) when what you have highlighted @dorpond appears to be a direct copy of the [token method for areas of effect on a grid](https://5e.tools/book.html#xge,2,token%20method,0). They are different both in approach and in results...
